### DISCUSSION TITLE
Redis Server integration with ApolloServer3

### DISCUSSION THREAD

**Discussion by jkshobby** [apollographql/apollo-server#6081] [Q&A] [answered]
Hi Team,
Am trying to do a POC where we wanted to integrate Redis cache with apolloserver. The document provided does not help in connecting all the wires.
in the Index.js I have the local redis server configured.

const server = new ApolloServer({
    typeDefs,
    resolvers,
    cache: new BaseRedisCache({
      client: new Redis({
        host: '127.0.0.1',
        port: '6379'
      })
    }),
  });
In the Resolver, trying to access the client so that I can set/get the KEY/Value to the redis instance.
Below code does not work as expected. Getting error since client cannot be accessed.

 Query: {
      get: (parent, { key }, { client }) => {
        try {
          return client.getAsync(key);
        } catch (e) {
          return null;
        }
      }
    },

  Mutation: {
    set: async (parent, { key, value }, context) => {
      try {
          console.log(key + value);
        await client.set(key, value);
        return true;
      } catch (e) {
        console.log(e);
        return false;
      }
    }
  }
Here is the typeDefs for this.
type Query {
get(key: String!): String
}

type Mutation {
set(key: String!, value: String!): Boolean!
}

Any help is really appreciate on this.

**Comment by glasser** [comment, marked as answer]
Passing `cache` to `new ApolloServer` makes the cache available for Apollo Server's own features (implemented in core or in plugins) that want to use a cache, such as the full operation response cache, the HTTP cache for the REST Data Source, the Automatic Persisted Query cache, etc. We have an interface `KeyValueCache` that we use to allow multiple implementations to work for these features (though we are planning to replace this with another project's abstraction, see #6045).

This cache isn't made directly available to your resolvers. The "context" object on which you are looking for a `client` is actually completely controlled by the `context` function you pass to your ApolloServer constructor. It's theoretically possible for you to pull the `cache` object out of Apollo Server and put it on `context` (though it's a bit awkward: your `context` function doesn't actually have an argument containing the cache, but you could write a little custom plugin whose requestDidStart copies `requestContext.cache` into `requestContext.context.cache` — confusing names, I know!).

However I wouldn't recommend you actually do that. The whole point of the abstract cache interface is to let Apollo Server's features work with any backend. But if you know you are using Redis, you don't need an abstraction. I suggest that you just use your favorite Redis client (eg https://www.npmjs.com/package/ioredis), create an ioredis client at startup, and put it on your `context` yourself. Then you can use the full Redis API instead of just the subset that is wrapped by our abstraction.

Something like:

```
import Redis from 'ioredis';

const redisClient = new Redis(...);

const server = new ApolloServer({
  typeDefs,
  resolvers,
  context: () => ({ redisClient }),
});
```

See https://www.apollographql.com/docs/apollo-server/data/resolvers/#the-context-argument for more details on the context function.

**Reply by jkshobby** [reply]
@glasser This is very helpful and was able to resolve my issue. Thank you so much.
